File: python/LeetCode/00744.py
from typing import List

def nextGreatestLetter(self, letters: List[str], target: str) -> str:
    if target >= letters[-1]:
        return letters[0]

    n = len(letters)

    low = 0
    high = n - 1
    while low <= high:
        mid = (low + high) // 2
        if letters[mid] <= target:
            low = mid + 1
        else:
            high = mid - 1

    return letters[low]

File: go/leetcode/00744.go
package main

func nextGreatestLetter(letters []byte, target byte) byte {
	if letters[len(letters)-1] <= target {
		return letters[0]
	}

	n := len(letters)
	high := n - 1
	low := 0

	for low <= high {
		mid := (low + high) / 2
		if letters[mid] <= target {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}

	return letters[low]
}
